### ISSUE TITLE
Allow DataLoader instead of Dataset to improve performance

### ISSUE THREAD

**Issue by FrancescoSaverioZuppichini** [dnouri/skorch#336]
Hi,

First of all, amazing work!

I have noticed that the train is really slow compared to a naive implementation with PyTorch. I am doing image segmentation, and with my dataset (4k image  128x128) with a 'vanilla' code I do one epoch every  30 seconds, while with skorch I do one epoch every  37 seconds!

I think the user should be able pre-defined `Dataloader`.  This will also allows a better integration with existing code since usually people are used to first define their Dataloader.

Also, in the `NeuralNet` constructor, there is a param `iterator_train` but it takes a constructor so I hacked it and I changed this line

```
# net.py line 1111
return  iterator(dataset, **kwargs)
```

to 

```
return  iterator
```

This allows to directly pass a custom dataloader to the model and improve performance, I now get the same time.

```
ds = TensorDataset(...)
dl = Datoader(ds, ...)

net = NeuralNet(iterator_train=dl)

net.fit(ds) # I still have to pass something here
```

So what if we had the possibility to directly pass a `Dataloader` in the fit method?

**Comment by taketwo** [comment]
I'm not sure I entirely understand your problem/proposal, but maybe this information can be relevant. It's possible to pass `DataLoader` parameters (such as `num_workers`) to the `NeuralNet` constructor:

```python
net = NeuralNet(
    iterator_train__num_workers=4,
    iterator_train__pin_memory=True
)
```

**Comment by benjamin-work** [comment]
> so I hacked it and I changed this line...

This is certainly no 'hack' :) We intended `NeuralNet` to be easily subclassed, so that overriding `get_*` methods such as `get_iterator` should be as easy as possible.

That said, your change could be flawed because you don't always want to return the same `DataLoader` from `get_iterator`. Say you want to train on some data and then predict on other data: with your change, you get back the same data for both. 

I wonder how the change you mentioned can make such a big difference. Could you maybe post some code that shows what iterator and dataset you are using? The main difference is that before your change, the `DataLoader` is initialized once per epoch, and after, only once per fit call. But initializing the `DataLoader` should not take much time.

**Comment by FrancescoSaverioZuppichini** [comment]
@taketwo Thank you to show how to actually pass the parameter. But, is this, in your opinion a little ugly? Would it be better to have the ability to pass your DataLoader? 

@benjamin-work It was slower since I was not able to pass the correct `num_workers` now everything is perfect

Also, what If I want to change the DataLoader at run time? I should create a new `NeuralNet`, right?

Thank you for the fast reply :)

**Comment by BenjaminBossan** [comment]
> But, is this, in your opinion a little ugly?

I don't think it's ugly, we borrowed this from sklearn and use it consistently to pass "deep" parameters around.

> It was slower since I was not able to pass the correct num_workers now everything is perfect

Great, otherwise something would have been wrong.

> Also, what If I want to change the DataLoader at run time? I should create a new NeuralNet, right?

No, there should be no need to create a new `NeuralNet`. So I believe there are two things that need explaining. 

First sklearn (and as an extension skorch) has the rule that your data should never be an attribute of the model, you should only pass it to the model via the correct methods like fit and predict. One advantage is, for example, that when you save your model, you don't save all your data as well (resulting in a potentially huge model file).

Second, I think of the `DataLoader` more as a way to access your data, not the proper data container itself (that would be the `Dataset`). Therefore, I would not initialize the net with an initialized `DataLoader` but rather only create it once it's needed.

All of this then has the advantage that you can easily call your model with different data:

```
net.fit(X_train, y_train)
net.predict_proba(X_valid)
net.predict(X_test)
```

**Comment by taketwo** [comment]
Great answer! Just a small addendum to address this question:

> Also, what If I want to change the DataLoader at run time? I should create a new NeuralNet, right?

You can do the following at "run time", after the model was instantiated:

```python
net.set_params(iterator_train__num_workers=4)
```

**Comment by FrancescoSaverioZuppichini** [comment]
Thank you
